File: GW3/LogFileHandling/Query/Tokens/TokenGroup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenGroup : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();

            return parser.PeekString().ToLower() == "group";
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            parser.NextString();
            if (parser.PeekString().ToLower() == "by")
                parser.NextString();
            return new TokenGroup();
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenGroup.ts
class TokenGroup extends Token {

    public tokenType: TokenType = TokenType.TokenGroup;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[{ suggestion: "group by" }];

    public canBeUsed(parser: QueryParser): boolean {
        parser.skipSpaces();
        return parser.peekString().toLowerCase() == "group";
    }

    public extract(parser: QueryParser): Token {
        parser.skipSpaces();
        parser.nextString();
        parser.skipSpaces();
        let token = new TokenGroup();
        token.value = "group";
        if (parser.peekString().toLowerCase() == "by") {
            token.value += " " + parser.nextString();
        }
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[] {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined') {
            switch (nextToken.tokenType) {
                case TokenType.TokenName:
                    return [
                        new TokenAscending(),
                        new TokenDescending(),
                        new TokenGroup(),
                        new TokenLimit(),
                        new TokenOrder(),
                        new TokenWhere()
                    ];
            }
        }
        return [];
    }
}